docs/decisions/0052-python-ai-connector-new-abstract-methods.md

```diff
@@ -0,0 +1,80 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: { accepted }
+contact: { Tao Chen }
+date: { 2024-09-03 }
+deciders: { Eduard van Valkenburg, Ben Thomas }
+consulted: { Eduard van Valkenburg }
+informed: { Eduard van Valkenburg, Ben Thomas }
+---
+
+# New abstract methods in `ChatCompletionClientBase` and `TextCompletionClientBase` (Semantic Kernel Python)
+
+## Context and Problem Statement
+
+The ChatCompletionClientBase class currently contains two abstract methods, namely `get_chat_message_contents` and `get_streaming_chat_message_contents`. These methods offer standardized interfaces for clients to engage with various models.
+
+> We will focus on `ChatCompletionClientBase` in this ADR but `TextCompletionClientBase` will be having a similar structure.
+
+With the introduction of function calling to many models, Semantic Kernel has implemented an amazing feature known as `auto function invocation`. This feature relieves developers from the burden of manually invoking the functions requested by the models, making the development process much smoother.
+
+Auto function invocation can cause a side effect where a single call to get_chat_message_contents or get_streaming_chat_message_contents may result in multiple calls to the model. However, this presents an excellent opportunity for us to introduce another layer of abstraction that is solely responsible for making a single call to the model.
+
+## Benefits
+
+- To simplify the implementation, we can include a default implementation of `get_chat_message_contents` and `get_streaming_chat_message_contents`.
+- We can introduce common interfaces for tracing individual model calls, which can improve the overall monitoring and management of the system.
+- By introducing this layer of abstraction, it becomes more efficient to add new AI connectors to the system.
+
+## Details
+
+### Two new abstract methods
+
+```python
+@abstractmethod
+async def _send_chat_request(
+    self,
+    chat_history: ChatHistory,
+    settings: PromptExecutionSettings
+) -> list[ChatMessageContent]:
+    pass
+```
+
+```python
+@abstractmethod
+async def _send_streaming_chat_request(
+    self,
+    chat_history: ChatHistory,
+    settings: PromptExecutionSettings
+) -> AsyncGenerator[list[StreamingChatMessageContent], Any]:
+    pass
+```
+
+### A new `ClassVar[bool]` variable in `ChatCompletionClientBase` to indicate whether a connector supports function calling
+
+This class variable will be overridden in derived classes and be used in the default implementations of `get_chat_message_contents` and `get_streaming_chat_message_contents`.
+
+```python
+class ChatCompletionClientBase(AIServiceClientBase, ABC):
+    """Base class for chat completion AI services."""
+
+    FUNCTION_CALLING_SUPPORT: ClassVar[bool] = False
```

Comment:
nit: name it as a question that's answered by the bool e.g. SUPPORTS_FUNCTION_CALLING, HAS_FUNCTION_CALLING, etc.